### PR TITLE
build with mkl-dnn by default

### PR DESCRIPTION
**Pull request by gujinghui**
build with mkl-dnn by default

### CODE REVIEW

**Comment by soumith** [comment]
```
[E ideep_operator.h:62] IDEEP error:could not initialize a memory descriptor
```

```
Oct 30 09:25:58 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/ccfAyF9m.s: Assembler messages:\n/tmp/ccfAyF9m.s:424: Error: no such instruction: `vextracti32x8 $0x1,%zmm0,%ymm0\'\n/tmp/ccfAyF9m.s:426: Error: no such instruction: `vpmullq %zmm0,%zmm1,%zmm0\'\n/tmp/ccfAyF9m.s:427: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:431: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm0\'\n/tmp/ccfAyF9m.s:434: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm2\'\n/tmp/ccfAyF9m.s:437: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:552: Error: no such instruction: `vextracti32x8 $0x1,%zmm0,%ymm1\'\n/tmp/ccfAyF9m.s:555: Error: no such instruction: `vpmullq %zmm0,%zmm1,%zmm0\'\n/tmp/ccfAyF9m.s:556: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:562: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm2\'\n/tmp/ccfAyF9m.s:565: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm2\'\n/tmp/ccfAyF9m.s:568: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm0\'\n/tmp/ccfAyF9m.s:679: Error: no such instruction: `vextracti32x8 $0x1,%zmm0,%ymm1\'\n/tmp/ccfAyF9m.s:682: Error: no such instruction: `vpmullq %zmm0,%zmm1,%zmm0\'\n/tmp/ccfAyF9m.s:683: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:688: Error: no such instruction: `vpmullq %zmm2,%zmm1,%zmm1\'\n/tmp/ccfAyF9m.s:690: Error: no such instruction: `vpmullq %zmm1,%zmm0,%zmm1\'\n/tmp/ccfAyF9m.s:693: Error: no such instruction: `vpmullq %zmm1,%zmm0,%zmm0\'\n" }
```

```
Oct 30 09:25:47 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/ccjDDAUX.s: Assembler messages:\n/tmp/ccjDDAUX.s:598: Error: operand size mismatch for `vmovdqu64\'\n/tmp/ccjDDAUX.s:621: Error: operand size mismatch for `vmovdqu64\'\n/tmp/ccjDDAUX.s:1157: Error: operand size mismatch for `vmovdqu64\'\n/tmp/ccjDDAUX.s:3545: Error: operand type mismatch for `vxorps\'\n/tmp/ccjDDAUX.s:3560: Error: operand type mismatch for `vxorps\'\n/tmp/ccjDDAUX.s:5656: Error: no such instruction: `vpermi2w %zmm17,%zmm15,%zmm18\'\n/tmp/ccjDDAUX.s:5671: Error: no such instruction: `vpmovwb %zmm18,%ymm3\'\n/tmp/ccjDDAUX.s:5672: Error: no such instruction: `vpermi2w %zmm19,%zmm16,%zmm4\'\n/tmp/ccjDDAUX.s:5673: Error: no such instruction: `vpmovwb %zmm4,%ymm4\'\n/tmp/ccjDDAUX.s:6736: Error: no such instruction: `vpermi2w %zmm6,%zmm4,%zmm2\'\n/tmp/ccjDDAUX.s:6741: Error: no such instruction: `vpmovwb %zmm2,%ymm2\'\n/tmp/ccjDDAUX.s:6754: Error: no such instruction: `vpermi2w %zmm7,%zmm5,%zmm16\'\n/tmp/ccjDDAUX.s:6755: Error: no such instruction: `vpmovwb %zmm16,%ymm1\'\n/tmp/ccjDDAUX.s:7899: Error: operand size mismatch for `vmovdqu64\'\n/tmp/ccjDDAUX.s:7916: Error: operand size mismatch for `vmovdqu64\'\n/tmp/ccjDDAUX.s:10975: Error: operand size mismatch for `vmovdqu64\'\n/tmp/ccjDDAUX.s:11530: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccjDDAUX.s:11614: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccjDDAUX.s:11722: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccjDDAUX.s:11803: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccjDDAUX.s:15557: Error: operand size mismatch for `vmovdqa64\'\n" }
Oct 30 09:25:47 
```


many other failures across CI. Maybe iDeep is not ready for default enabling yet?

**Comment by gujinghui** [comment]
@soumith 
Trying to root-cause these issues. :)

For #1, should be caused by the ideep case, copy_op_test. Will fix it soon.

For other compiling issues, according to previous study, should be caused by building environment when building mkl-dnn, for example, the version of binutils. Could you tell who can help me on such issue?

**Comment by gujinghui** [comment]
@soumith 
Thanks for reply.

The cases with below failure have been modified here.
> [E ideep_operator.h:62] IDEEP error:could not initialize a memory descriptor

The reason is zero dim is still not supported for now both in ideep and mkl-dnn.
I have to skip such cases to avoid failures now.

But the coming upgrade will introduce zero dim feature soon.
The cases will be enabled after next upgrade.

Besides, we still have rest 3 kinds of failures in this PR:
1.  illegal instructions in compiling or cases.
> ```
> Oct 30 09:25:58 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/ccfAyF9m.s: Assembler messages:\n/tmp/ccfAyF9m.s:424: Error: no such instruction: `vextracti32x8 $0x1,%zmm0,%ymm0\'\n/tmp/ccfAyF9m.s:426: Error: no such instruction: `vpmullq %zmm0,%zmm1,%zmm0\'\n/tmp/ccfAyF9m.s:427: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:431: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm0\'\n/tmp/ccfAyF9m.s:434: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm2\'\n/tmp/ccfAyF9m.s:437: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:552: Error: no such instruction: `vextracti32x8 $0x1,%zmm0,%ymm1\'\n/tmp/ccfAyF9m.s:555: Error: no such instruction: `vpmullq %zmm0,%zmm1,%zmm0\'\n/tmp/ccfAyF9m.s:556: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:562: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm2\'\n/tmp/ccfAyF9m.s:565: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm2\'\n/tmp/ccfAyF9m.s:568: Error: no such instruction: `vpmullq %zmm2,%zmm0,%zmm0\'\n/tmp/ccfAyF9m.s:679: Error: no such instruction: `vextracti32x8 $0x1,%zmm0,%ymm1\'\n/tmp/ccfAyF9m.s:682: Error: no such instruction: `vpmullq %zmm0,%zmm1,%zmm0\'\n/tmp/ccfAyF9m.s:683: Error: no such instruction: `vpmullq %zmm0,%zmm2,%zmm2\'\n/tmp/ccfAyF9m.s:688: Error: no such instruction: `vpmullq %zmm2,%zmm1,%zmm1\'\n/tmp/ccfAyF9m.s:690: Error: no such instruction: `vpmullq %zmm1,%zmm0,%zmm1\'\n/tmp/ccfAyF9m.s:693: Error: no such instruction: `vpmullq %zmm1,%zmm0,%zmm0\'\n" }
> ```
I'm trying to repro this locally. But still suspect this should be building environment issue.
If got solid evident, will return to you soon.

2. not find libmkldnn.dylib on macOS.
Perhaps, it should be FindMKLDNN.cmake issue, which gives lib path for libmkldnn.
The lib path should be corrected for macOS.

3. crashed due to wrong deconstructor of mkldnn_engine
Under debuging. If got any update, will send here.

**Comment by kleisauke** [comment]
FYI: Building PyTorch (using CMake's `ExternalProject`)  with the `-DUSE_MKLDNN=ON` flag causes linkage errors for me (after deleting the build directory):
```bash
make[2]: *** No rule to make target 'third_party/pytorch/src/pytorch-build/lib/libmkldnn.so', needed by '../lib/libtest.so.5.0.0'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:109: src/api/CMakeFiles/test.dir/all] Error 2
make: *** [Makefile:95: all] Error 2
```

It seems that `libmkldnn.so` is installed into the build folder (for e.g. `/home/pytorch-test/build/third_party/pytorch/src/pytorch-build/lib/libmkldnn.so`), see:
https://github.com/pytorch/pytorch/blob/8444ed951dfeae934f98ce8165e9ac40dbbadb85/cmake/Modules/FindMKLDNN.cmake#L106

Maybe MKL-DNN needs to be installed in `CMAKE_INSTALL_PREFIX`?

**Comment by gujinghui** [comment]
> FYI: Building PyTorch (using CMake's `ExternalProject`) with the `-DUSE_MKLDNN=ON` flag causes linkage errors for me (after deleting the build directory):
> 
> ```shell
> make[2]: *** No rule to make target 'third_party/pytorch/src/pytorch-build/lib/libmkldnn.so', needed by '../lib/libtest.so.5.0.0'.  Stop.
> make[1]: *** [CMakeFiles/Makefile2:109: src/api/CMakeFiles/test.dir/all] Error 2
> make: *** [Makefile:95: all] Error 2
> ```
> 
> It seems that `libmkldnn.so` is installed into the build folder (for e.g. `/home/pytorch-test/build/third_party/pytorch/src/pytorch-build/lib/libmkldnn.so`), see:
> https://github.com/pytorch/pytorch/blob/8444ed951dfeae934f98ce8165e9ac40dbbadb85/cmake/Modules/FindMKLDNN.cmake#L106
> 
> Maybe MKL-DNN needs to be installed in `CMAKE_INSTALL_PREFIX`?

what's your build command? do you have full build log?

**Comment by gujinghui** [comment]
rebased with latest code base.

Finally, we passed all test cases. 🎉 🎉 🎉 
@yinghai @soumith 
Please help review.

@Jianhui-Li @jgong5

@kleisauke This PR should be able to fix the issue in your code.
Please have a try.

**Comment by kleisauke** [comment]
@gujinghui I can confirm that this PR resolves the issue that I had, thanks!

**Comment by yinghai** [comment]
Thanks, @gujinghui!. @orionr, do we have a CI setup to test this?

**Comment by orionr** [comment]
@yinghai doesn't look like we have OSS CI with MKLDNN enabled - likely want to test that. However, we do have internal Sandcastle testing of it.

If you want to add it, feel free to tweak .jenkins/pytorch/build.sh with something like

```
if ! which conda; then
  pip install -q mkl mkl-devel
  export USE_MKLDNN=1
fi
```

**Comment by orionr** [comment]
@gujinghui let me push a change to the CI so MKLDNN gets tested.

**Comment by orionr** [comment]
Changes pushed. Let's see how CI does.

**Comment by avmgithub** [comment]
ppc64le does not support mkl-dnn.  I suspect this is why the ppc64le is now breaking.


--   USE_MKLDNN            : ON   <<<<<<<<<<<<<<<<<


c++: error: unrecognized command line option '-march=native'
third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/build.make:62: recipe for target 'third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/batch_normalization.cpp.o' failed
make[2]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/batch_normalization.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
c++: error: unrecognized command line option '-march=native'
third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/build.make:75: recipe for target 'third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/convolution.cpp.o' failed
make[2]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/convolution.cpp.o] Error 1
CMakeFiles/Makefile2:1342: recipe for target 'third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/all' failed
make[1]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

 


Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Failed to run 'bash ../tools/build_pytorch_libs.sh --use-cuda --use-fbgemm --use-nnpack --use-mkldnn --use-qnnpack caffe2'

**Comment by gujinghui** [comment]
> ppc64le does not support mkl-dnn. I suspect this is why the ppc64le is now breaking.
> 
> -- USE_MKLDNN : ON <<<<<<<<<<<<<<<<<
> 
> c++: error: unrecognized command line option '-march=native'
> third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/build.make:62: recipe for target 'third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/batch_normalization.cpp.o' failed
> make[2]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/batch_normalization.cpp.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> c++: error: unrecognized command line option '-march=native'
> third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/build.make:75: recipe for target 'third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/convolution.cpp.o' failed
> make[2]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/common/convolution.cpp.o] Error 1
> CMakeFiles/Makefile2:1342: recipe for target 'third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/all' failed
> make[1]: *** [third_party/ideep/mkl-dnn/src/CMakeFiles/mkldnn.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> 
> Makefile:140: recipe for target 'all' failed
> make: *** [all] Error 2
> Failed to run 'bash ../tools/build_pytorch_libs.sh --use-cuda --use-fbgemm --use-nnpack --use-mkldnn --use-qnnpack caffe2'

Please try to disable MKLDNN by settting NO_MKLDNN=1:
NO_MKLDNN=1 python setup.py build_deps
